druid-shell/src/platform/x11/xkb.rs

```diff
@@ -0,0 +1,456 @@
+        };
+        if id != -1 {
+            Some(DeviceId(id))
+        } else {
+            None
+        }
+    }
+
+    pub fn keymap_from_device(&self, conn: &XCBConnection, device: DeviceId) -> Option<Keymap> {
+        let key_map = unsafe {
+            xkb_x11_keymap_new_from_device(
+                self.0,
+                conn.get_raw_xcb_connection() as *mut xcb_connection_t,
+                device.0,
+                XKB_KEYMAP_COMPILE_NO_FLAGS,
+            )
+        };
+        if key_map.is_null() {
+            return None;
+        }
+        Some(Keymap(key_map))
+    }
+
+    /// Set the log level using `log` levels.
+    ///
+    /// Becuase `xkb` has a `critical` error, each rust error maps to 1 above (e.g. error ->
+    /// critical, warn -> error etc.)
+    pub fn set_log_level(&self, level: tracing::Level) {
+        use tracing::Level;
+        let level = match level {
+            Level::ERROR => XKB_LOG_LEVEL_CRITICAL,
+            Level::WARN => XKB_LOG_LEVEL_ERROR,
+            Level::INFO => XKB_LOG_LEVEL_WARNING,
+            Level::DEBUG => XKB_LOG_LEVEL_INFO,
+            Level::TRACE => XKB_LOG_LEVEL_DEBUG,
+        };
+        unsafe {
+            xkb_context_set_log_level(self.0, level);
+        }
+    }
+}
+
+impl Clone for Context {
+    fn clone(&self) -> Self {
+        Self(unsafe { xkb_context_ref(self.0) })
+    }
+}
+
+impl Drop for Context {
+    fn drop(&mut self) {
+        unsafe {
+            xkb_context_unref(self.0);
+        }
+    }
+}
+
+pub struct Keymap(*mut xkb_keymap);
+
+impl Keymap {
+    pub fn state(&self) -> State {
+        State::new(self)
+    }
+}
+
+impl Clone for Keymap {
+    fn clone(&self) -> Self {
+        Self(unsafe { xkb_keymap_ref(self.0) })
+    }
+}
+
+impl Drop for Keymap {
+    fn drop(&mut self) {
+        unsafe {
+            xkb_keymap_unref(self.0);
+        }
+    }
+}
+
+pub struct State {
+    state: *mut xkb_state,
+    mods: ModsIndices,
+}
+
+#[derive(Clone, Copy)]
+pub struct ModsIndices {
+    control: xkb_mod_index_t,
+    shift: xkb_mod_index_t,
+    alt: xkb_mod_index_t,
+    super_: xkb_mod_index_t,
+    caps_lock: xkb_mod_index_t,
+    num_lock: xkb_mod_index_t,
+}
+
+impl State {
+    pub fn new(keymap: &Keymap) -> Self {
+        let keymap = keymap.0;
+        let state = unsafe { xkb_state_new(keymap) };
+        let mod_idx = |str: &'static [u8]| unsafe {
+            xkb_keymap_mod_get_index(keymap, str.as_ptr() as *mut c_char)
+        };
+        Self {
+            state,
+            mods: ModsIndices {
+                control: mod_idx(XKB_MOD_NAME_CTRL),
+                shift: mod_idx(XKB_MOD_NAME_SHIFT),
+                alt: mod_idx(XKB_MOD_NAME_ALT),
+                super_: mod_idx(XKB_MOD_NAME_LOGO),
+                caps_lock: mod_idx(XKB_MOD_NAME_CAPS),
+                num_lock: mod_idx(XKB_MOD_NAME_NUM),
+            },
+        }
+    }
+
+    pub fn key_event(&self, scancode: u32, state: KeyState) -> KeyEvent {
```

Comment:
This feels like it should be `&mut self`

Reply:
again a bit unsure, given we only work with `*mut`s and this is not `Sync/Send`. this can be thought of as interior mutability

BUT, the `&mut` doesn't hurt us in this case, lets do it anyways